pythonx/UltiSnips/remote_pdb.py

```diff
@@ -0,0 +1,109 @@
+import sys
+from bdb import BdbQuit
+
+from UltiSnips import vim_helper
+import threading
+
+class RemotePDB(object):
+    """
+    Launch a pdb instance listening on (host, port).
+    Used to provide debug facilities you can access with netcat or telnet.
+    """
+    
+    singleton = None
+
+    def __init__(self, host, port):
+        self.host = host
+        self.port = port
+        self._pdb = None
+
+    def start_server(self):
+        """
+        Create an instance of Pdb bound to a socket
+        """
+        if self._pdb is not None :
+          return
+        import pdb 
+        import socket
+
+        self.server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, True)
+        self.server.bind((self.host, self.port))
+        self.server.listen(1)
+        self.connection, address = self.server.accept()
+        io = self.connection.makefile("rw")
+        parent = self
+        
+        class Pdb(pdb.Pdb):
+            """Patch quit to close the connection"""
+            def set_quit(self):
+                parent._shutdown()
+                super().set_quit()
+                
+        self._pdb = Pdb(stdin=io, stdout=io)
+
+    def _pm(self, tb):
+        """
+        Launch the server as post mortem on the currently handled exception
+        """
+        try:
+            self._pdb.interaction(None, tb) 
+        except: # Ignore all exceptions part of debugger shutdown (and bugs... https://bugs.python.org/issue44461 )
+            pass
+
+    def set_trace(self, frame):
+        self._pdb.set_trace(frame)
+
+    def _shutdown(self):
+        if self._pdb is not None :
+            import socket
+            self.connection.shutdown(socket.SHUT_RDWR)
+            self.connection.close()
+            self.server.close()
+            self._pdb = None
+
+    @classmethod
```

Comment:
Some of these @classmethods are not using `cls`, hence they should be stand alone methods in this module for minimized coupling.

Reply:
It is to permit subclassing, and also to to group all the RemotePdb related method to the same class. This way, when other methods call them, we know immediately we just have to look in the same class. Though, they could be made staticmethod

Reply:
+1 for staticmethods for sure, however I still think they should be standalone: I respect the argument that they are grouped through the class. But then again, they are also grouped through the module name which is fairly descriptive. Making them stand alone strongly communicates that they do not need to know anything about the class or its internal setup, also reducing the API surface of the class making it more granular and hence easier to change in the future.

I do not feel super strongly, I can live with staticmethods, but I'd prefer them to be standalone.